cpp/src/arrow/io/file.cc

```diff
@@ -356,49 +360,63 @@ class MemoryMappedFile::MemoryMap : public MutableBuffer {
     }
   }
 
-  Status Open(const std::string& path, FileMode::type mode) {
-    int prot_flags;
-    int map_mode;
-
+  Status Open(const std::string& path, FileMode::type mode, int64_t size = -1) {
     file_.reset(new OSFile());
 
     if (mode != FileMode::READ) {
       // Memory mapping has permission failures if PROT_READ not set
-      prot_flags = PROT_READ | PROT_WRITE;
-      map_mode = MAP_SHARED;
+      prot_flags_ = PROT_READ | PROT_WRITE;
+      map_mode_ = MAP_SHARED;
       constexpr bool append = false;
       constexpr bool truncate = false;
       constexpr bool write_only = false;
       RETURN_NOT_OK(file_->OpenWriteable(path, truncate, append, write_only));
 
       is_mutable_ = true;
     } else {
-      prot_flags = PROT_READ;
-      map_mode = MAP_PRIVATE;  // Changes are not to be committed back to the file
+      prot_flags_ = PROT_READ;
+      map_mode_ = MAP_PRIVATE;  // Changes are not to be committed back to the file
       RETURN_NOT_OK(file_->OpenReadable(path));
 
       is_mutable_ = false;
     }
 
-    size_ = file_->size();
+    // Memory mapping fails when file size is 0, if it is 0,
+    // delay it until the first resize, if it
+    if (file_->size() > 0) {
+      RETURN_NOT_OK(InitMMap(file_->size()));
+    }
 
-    void* result = nullptr;
+    position_ = 0;
 
-    // Memory mapping fails when file size is 0
-    if (size_ > 0) {
-      result =
-          mmap(nullptr, static_cast<size_t>(size_), prot_flags, map_mode, file_->fd(), 0);
-      if (result == MAP_FAILED) {
-        std::stringstream ss;
-        ss << "Memory mapping file failed: " << std::strerror(errno);
-        return Status::IOError(ss.str());
+    return Status::OK();
+  }
+
+  // Sets the size to the specified value. If the current position
+  // is further then the new size, sets the position to the last byte.
+  Status Resize(const int64_t new_size, bool shrink_to_fit = false) {
+    if (!shrink_to_fit || (new_size > capacity_)) {
+      RETURN_NOT_OK(Reserve(new_size));
```

Comment:
Here is as good a place as any to explain some design decisions. The `MemoryMappedFile::MemoryMap` was a subclass of `MutableBuffer` initially. I changed it to `ResizableBuffer` and tried to keep the abstraction. So a lot of logic tries to mirror the logic of the `PoolBuffer`. This explains the `shrink_to_fit` check, the differences between `size_` and `capacity_` and the memory aligning (since the buffer guarantee a 64-bit alignment). 

On the other hand, I tried to hide these details from the `MemoryMappedFile` interface, that's why I return `capacity_` as size, as it is the actual physical size of the file.

Now, I not entirely convinced either that this two-level abstraction makes sense, especially since `MemoryMappedFile::MemoryMap` isn't even available publicly. A couple of options I see here:

1) Keep the hierarchy, simply adjust `MemoryMap` - the most conservative option in my opinion, since it keeps the original abstractions and structure intact.

2) Lose `MemoryMap` and implement the functionality directly in `MemoryMappedFile`. Will simplify some logic and not a lot of the `Buffer` interface of the `memory_map_` is used anyway, just `data()`, and the need for stuff like `position()` and `advance()` suggest a leaky abstraction anyway.

3) Promote `MemoryMap` to a fist level buffer (rename it `MemoryMapBuffer`) and expose it to the user. Would increase the interface size and would probably require a lot more tests. The most  radical option, so I'm not sure about it.

We can move this discussion to the mailing list or JIRA, if you think it's better (not sure how these things are handled usually).

Reply:
I would go for option 2 ("Lose `MemoryMap` and implement the functionality directly in `MemoryMappedFile`"). Or perhaps option 2b ("Lose `MemoryMap` and implement the functionality in a `MemoryMappedFile::Impl` class that doesn't inherit from `Buffer`").

Reply:
Somewhere you need a subclass of `Buffer` to "own" the memory map, though. The buffers that are read out of a `MemoryMappedFile` will often exceed the lifetime of `MemoryMappedFile`, so the purpose of `MemoryMap` is to extend the lifetime of the memory map until all references to it have been destroyed.

Reply:
Right... then we have a problem, because resizing the `MemoryMap` will invalidate previously exported buffers (`mremap` doesn't necessarily reuse the old address space). We can instead create a new memory mapping, but this can lead to consuming a lot of address space since parts of the file will be mapped twice or more (I'm also unsure about performance).

This hints that perhaps the random-access `MemoryMappedFile` should not be resizable at all, and instead we could have another append-only `MemoryMappedOutputStream` that would resize transparently (assuming performance benefits make the addition worthwhile, which I'm a bit skeptical about).

Side question: generally speaking, do we ever need to open files in read+write mode?

Reply:
Yeah, I would agree that it would be bad to invalidate buffers in a read+write `MemoryMappedFile`.

I'm also skeptical about the performance benefits here. 

As an aside, it seems like more benefits could be reaped by creating a write queue / "spooler" abstraction to keep the disk busy while new bits of data are being prepared to write, performing writes in a background thread. We don't have much in the way of asynchronous IO yet (background reads / writes, background IPC streaming); it would be useful IMHO to start defining some abstractions for async IO that fit in with the rest of the Arrow IO and memory management abstractions.

Reply:
This reminds me of [a similar discussion](http://mail-archives.apache.org/mod_mbox/arrow-dev/201804.mbox/%3CCACu5Zf%3DBa7tL4FjPigxpVJX2_3EmPsyc93Oiy-OwWbobzeKbKA%40mail.gmail.com%3E). Basically, it's hard to combine `ResizableBuffer` with `Slices`. Antoine proposed a solution similar to Python's `getbuf`, which increments an `exports` counter and prevents resizing if it's over 0. It would at make the resizing safe, as well as expand the number of use cases for `ResizableBuffer`.

I don't quite understand which performance benefits can be expected, i.e. compared to what? Closing a map, resizing the file and creating a new map as an alternative won't have a significant impact on the performance, since resizing is done comparatively rarely. With this change I aimed for API improvement, since the expectation that you can resize an (mmap'ed) file feel quite natural to me.

Reply:
So, we can look at the `use_count()` of the `shared_ptr` to determine whether any references to the memory map have been exported. It should be 1 unless someone has a reference to it. So in that case, return IOError. I think that should resolve that issue at least? We should run benchmarks comparing sequential writes with `FileOutputStream` vs. this.

Reply:
> it seems like more benefits could be reaped by creating a write queue / "spooler" abstraction to keep the disk busy while new bits of data are being prepared to write, performing writes in a background thread

I'm curious about this. Aren't writes generally asynchronous nowadays (i.e. when `write()` returns, the data typically isn't written out to disk yet, just stored somewhere in the page cache)?

Reply:
Let's consider a case where we have some code that looks like:

```
while (widget->HaveMoreWork()) {
  std::shared_ptr<Buffer> some_data = widget->GetWork();
  RETURN_NOT_OK(out_stream->Write(some_data));
}
```

Maybe we're only getting to 50% IO-bandwidth saturation. So we could have a background write-spooler where `Write` is non-blocking up to some queue-ing depth (or perhaps queueing controlled by some other limits, like size of a temporary memory pool, so you don't blow up your memory when the work to write piles up). The goal would be to make it easier to reach maximum write output.

There's some other considerations, like how to handle the juggling act of multiple chunks of scratch space. There's probably some libraries that deal with a lot of concurrency issues, like Facebook's Folly library, that we could learn some design patterns from

Reply:
I admit this pattern would be more useful / have more impact when reading and processing data from a file in some cases

Reply:
Regarding `use_count()`: we could use it, but, since we want it to be thread safe, we'd have to lock the reading first and hold the lock until we've checked the value of `use_count()`. As a consequence, we'd have to acquire a lock on every `ReadAt` call. A simple mutex here might be too expensive, and since we neither use C++17 nor `boost::thread`, we don't have a ready available implementation of a shared lock. A hand-made implementation can be done, of course, but a existing and tested implementation would save a lot of headache now and in the future.

So first I'd like to measure the performance impact of putting a mutex around `ReadAt` (the one with an output buffer). 

But regardless of the result: what's your opinion on including boost-threading, not necessarily for this change? boost being boost, it will also simplify the move to the newer C++ standards, compared to a hand-made solution.

Reply:
I'm not sure what a mutex would solve. The problem here is, if you call `Resize` after `ReadAt`, the buffer exported by `ReadAt` may become invalid. Note: this is after `ReadAt` has returned, so protecting `ReadAt` itself doesn't seem sufficient.

Reply:
Hi Antoine, holding the mutex, we would check use_count() for values
greater than 1 and fail if we detect them.

But it would still leave the problem of slices of slices. I really think this issue should be solven on the level of buffers, though.